agent/consul/state/acl.go

```diff
@@ -986,6 +1027,11 @@ func (s *Store) aclTokenGetTxn(tx *memdb.Txn, ws memdb.WatchSet, value, index st
 
 	if rawToken != nil {
 		token := rawToken.(*structs.ACLToken)
+
+		if !fixLinks {
+			return token, nil
+		}
+
```

Comment:
In the event publisher when a policy/role gets deleted we look up the subscriber tokens to see which ones are affected, and if the links get fixed in the process then we can't tell which ones had the policy/role that was deleted and need to be notified

Reply:
🤔 got it.

I'm not familiar enough with how and when this fixup process works to know if this is right/safe or not. Since we added the channel into event publish to make it Async, that now means that this read in response to an ACL DELETE event is not atomic with the DELETE itself - if there is some background process that is cleaning up the links for example it could be possible that that process picked up the delete even quicker and performed the actual fixup on the memDB snapshot that we are now querying.

In other words, whatever was broken before you added this, could still be broken the same way in a rare race which makes me think we should maybe have a different approach.

[edit]

Actually I was about to propose it would be safer if we pass through a snapshot on the `commitCh` so we know that any reads done during publish are effectively "atomic" with the state at the publish time but then I checked and you already do that.

So I *think* this is OK but still not 100% sure why it's the right thing to do given that fixing up links is probably something we should ideally not make async and if we _did_ do them inline it would be equivalent of this change not being present. Do we just need need to read the ACL token before the delete and somehow pass that through to the event rather than rely on us not getting the state store referentially consistent right after the delete?

Reply:
I think this makes sense. Also if you are not sending the retrieved token/role back to the user then there shouldn't be any need to fix the links at all. All of the internal ACL token resolution related things take into account linking to deleted things and nothing internally cares about the link's Name.

Reply:
@mkeeler so you're saying this change (skipping link fixing) shouldn't be necessary? I only added it after hitting test failures where looking up the token was removing the deleted policy/role from the links

Reply:
I think Matt was saying that your code here makes sense rather than my rambling pseudo concern @kyhavlov  😄